hw/xbox/mcpx_apu.c

```diff
@@ -811,11 +1128,346 @@ static const MemoryRegionOps ep_ops = {
+                value = 0.0f;
+        } else {
+            //FIXME: This formula and threshold is not accurate, but I can't get it any better for now
+            value = 255.0f * powf(0.99988799f, (decay_rate * 16 - count) * 4096 / decay_rate);
+        }
+        if (value <= (sustain_level + 0.2f) || (value > 255.0f)) {
+                //FIXME: Should we still update lvl?
+                cur++;
+                voice_set_mask(d, v, NV_PAVS_VOICE_PAR_STATE, cur_mask, cur);
+        } else {
+                count--;
+                voice_set_mask(d, v, NV_PAVS_VOICE_CUR_ECNT, count_mask, count);
+                voice_set_mask(d, v, lvl_reg, lvl_mask, value);
+        }
+        return value / 255.0f;
+    }
+    case 5: { // Sustain
+        uint8_t sustain_level = voice_get_mask(d, v, reg_a, NV_PAVS_VOICE_CFG_ENVA_EA_SUSTAINLEVEL);
+        voice_set_mask(d, v, NV_PAVS_VOICE_CUR_ECNT, count_mask, 0x00); // FIXME: is this only set to 0 once or forced to zero?
+        voice_set_mask(d, v, lvl_reg, lvl_mask, sustain_level);
+        return sustain_level / 255.0f;
+    }
+    case 6: { // Release
+        uint16_t release_rate = voice_get_mask(d, v, rr_reg, rr_mask);
+        uint16_t count = voice_get_mask(d, v, NV_PAVS_VOICE_CUR_ECNT, count_mask);
+        count--;
+        voice_set_mask(d, v, NV_PAVS_VOICE_CUR_ECNT, count_mask, count);
+        uint8_t lvl = voice_get_mask(d, v, lvl_reg, lvl_mask);
+        float value = count * lvl / (release_rate * 16);
+        if (count == 0) {
+            //FIXME: What to do now?!
+#if 0 // Hack so we don't assert
+            cur++; // FIXME: Does this happen?!
+            voice_set_mask(d, v, NV_PAVS_VOICE_PAR_STATE, cur_mask, cur);
+#else
+            voice_set_mask(d, v, NV_PAVS_VOICE_PAR_STATE, cur_mask, 0x0); // Is this correct? FIXME: Turn off voice?
+#endif
+        }
+        return value / 255.0f;
+    }
+    case 7: // Force release
+        assert(false); //FIXME: This mode is not understood yet
+        return 1.0f;
+    default:
+        fprintf(stderr, "Unknown envelope state 0x%x\n", cur);
+        assert(false);
+    }
+
+    return 0;
+}
+#endif
+
 static void process_voice(MCPXAPUState *d,
                           int32_t mixbins[NUM_MIXBINS][NUM_SAMPLES_PER_FRAME],
                           uint32_t voice)
 {
-    /* FIXME: Implement */
+#if 0
+    uint32_t v = voice;
+    int32_t samples[2][0x20] = {0};
+
+    float ea_value = step_envelope(d, v, NV_PAVS_VOICE_CFG_ENV0, NV_PAVS_VOICE_CFG_ENVA, NV_PAVS_VOICE_TAR_LFO_ENV, NV_PAVS_VOICE_TAR_LFO_ENV_EA_RELEASERATE, NV_PAVS_VOICE_PAR_OFFSET, NV_PAVS_VOICE_PAR_OFFSET_EALVL, NV_PAVS_VOICE_CUR_ECNT_EACOUNT, NV_PAVS_VOICE_PAR_STATE_EACUR);
+    float ef_value = step_envelope(d, v, NV_PAVS_VOICE_CFG_ENV1, NV_PAVS_VOICE_CFG_ENVF, NV_PAVS_VOICE_CFG_MISC, NV_PAVS_VOICE_CFG_MISC_EF_RELEASERATE, NV_PAVS_VOICE_PAR_NEXT, NV_PAVS_VOICE_PAR_NEXT_EFLVL, NV_PAVS_VOICE_CUR_ECNT_EFCOUNT, NV_PAVS_VOICE_PAR_STATE_EFCUR);
```

Comment:
Such stuff, which affects timing, might need shadow-registers if we want to synchronize VP to in realtime. We'd basically need a set of shadow registers for playback (unreachable from Xbox; real-time), and a set of emulated registers (reachable from Xbox; virtual-time).